[DO NOT MERGE][Reproducing issue] : View converter , bs >1

SrivastavaKshitij opened the pull request:
Hi @jaybdub

I recently came across a use case where the conversion fails for view converter when batch size >1 . 

Command: `python -m torch2trt.test --name=view` 

Some pointers:

1. I have disabled all other tests except one to reproduce the error that I am seeing.
2. One interesting thing that I noticed was that even though I was using an input tensor size of `[2,3,3,3]` , view converter was still seeing the input size as `[1,3,3,3]` . Line 15 in the `view` converter. I couldnt de-bug it though.

jaybdub committed:
Merge pull request #56 from NVIDIA-AI-IOT/pyt1.2

removed caffe2 dependency

mt1871 committed:
consider ceil_mode of torch.nn.MaxPool2d

If ceil_mode is False, the default value of layer.padding_mode is
PaddingMode.EXPLICIT_ROUND_DOWN. If ceil_mode is True, padding_mode
should be trt.PaddingMode.EXPLICIT_ROUND_UP.

jaybdub committed:
registered add for more interfaces

jaybdub committed:
added additional div interfaces

jaybdub committed:
added additional mul interfaces

jaybdub committed:
Merge pull request #61 from NVIDIA-AI-IOT/torch_methods

Torch methods

jaybdub committed:
added clamp, clamp_min, clamp_max

jaybdub committed:
added clamp tensor

jaybdub committed:
added normalize

jaybdub committed:
Merge pull request #62 from NVIDIA-AI-IOT/normalize

clamp and normalize

mt1871 committed:
added test cases for MaxPool2d

jaybdub committed:
Merge pull request #57 from mt1871/fix_maxpool_ceilmode

consider ceil_mode of torch.nn.MaxPool2d

jaybdub committed:
added mobilenet_v2 to module tests

jaybdub committed:
Merge pull request #70 from NVIDIA-AI-IOT/mobilenet_v2

added mobilenet_v2 to module tests

jaybdub committed:
added parsed_args

jaybdub committed:
added utilities to support normalize

jaybdub committed:
dtype to normalized constants

jaybdub committed:
added avg_pool2d and removed AvgPool2d

jaybdub committed:
fixed normalize

jaybdub committed:
added missing avg_pool2d.py

jaybdub committed:
added max_pool2d removed MaxPool2d

jaybdub committed:
fixed arg parsing in avg_pool2d and max_pool2d

jaybdub committed:
refactor

jaybdub committed:
Merge pull request #73 from NVIDIA-AI-IOT/functional_refactor

- adds avg_pool2d
- adds max_pool2d
- removes AvgPool2d
- removes MaxPool2d
- adds get_arg(...)
- adds torch_dim_to_trt_axes(...)
- adds add_trt_constant(...)

jaybdub committed:
added split and chunk

jaybdub committed:
added shufflenet to module tests

jaybdub committed:
Merge pull request #74 from NVIDIA-AI-IOT/split

- adds ``torch.chunk`` and ``torch.Tensor.chunk``
- adds ``torch.split`` and ``torch.Tensor.split``
- adds tests for ``squeezenet*`` models

jaybdub committed:
added permute

jaybdub committed:
fixed permute converter name

jaybdub committed:
Merge pull request #75 from NVIDIA-AI-IOT/permute

added permute

jaybdub committed:
added trt_ method to get/create/broadcast tensors

jaybdub committed:
modified normalize to use trt_

jaybdub committed:
fixed trt_ to handle single tenosr outputs

jaybdub committed:
updated converters to use trt_ instead of input._trt

jaybdub committed:
added __radd__ to support left hand primitives

jaybdub committed:
added rdiv and radd

jaybdub committed:
added rdiv, rmul, radd

jaybdub committed:
replcaed input._trt with trt_ in interpolate

jaybdub committed:
Merge pull request #80 from NVIDIA-AI-IOT/prepare_broadcastable_tensors

Prepare broadcastable tensors

jaybdub committed:
added mnasnet to test

jaybdub committed:
Merge pull request #81 from NVIDIA-AI-IOT/prepare_broadcastable_tensors

added mnasnet to test

jaybdub committed:
refactored tests

jaybdub committed:
refactored tests

jaybdub committed:
Merge branch 'extra_tests' of https://github.com/NVIDIA-AI-IOT/torch2trt into extra_tests

jaybdub committed:
fix missing --include argument in test.py

jaybdub committed:
Merge pull request #84 from NVIDIA-AI-IOT/extra_tests

refactored tests

- adds --include flag to test.py to load additional tests
- moves torchivsion tests from torch2trt.module_test to torch2trt.tests.torchvision.classification
- adds torchvision segmentation tests to torch2trt.tests.torchvision.segmentation

jaybdub committed:
removed copy(..) from conversion hook

jaybdub committed:
removed copy(..) from conversion hook

jaybdub committed:
Merge branch 'py2_fix' of https://github.com/NVIDIA-AI-IOT/torch2trt into py2_fix

jaybdub committed:
handle __div__, __rdiv__ and __idiv__ to support Python2

jaybdub committed:
Merge pull request #11 from narendasan/narendasan/gitignore

Adds a gitignore

jaybdub committed:
Merge pull request #10 from narendasan/narendasan/typo_in_readme

Missing an import in the first example

jaybdub committed:
Merge pull request #85 from NVIDIA-AI-IOT/py2_fix

Py2 fix

jaybdub committed:
Merge branch 'master' into narendasan/custom_dep_locations

jaybdub committed:
removes caffe2 from build

caffe2 is not needed and breaks build when using PyTorch 1.2

jaybdub committed:
Merge pull request #12 from narendasan/narendasan/custom_dep_locations

Adds a way to specify custom library locations and remove build artifacts

jaybdub committed:
added support for prelu

jaybdub committed:
Merge pull request #88 from NVIDIA-AI-IOT/prelu

added support for prelu

jaybdub committed:
added supported unary ops

jaybdub committed:
Merge pull request #89 from NVIDIA-AI-IOT/unary_supported

added supported unary ops

jaybdub committed:
added max reduce/elementwise

jaybdub committed:
added min reduce/elementwise

jaybdub committed:
added sub

jaybdub committed:
added pow elementwise

jaybdub committed:
fixed min/max reduce all

jaybdub committed:
added prod

jaybdub committed:
added sum reduce

jaybdub committed:
increment patch

jaybdub committed:
separate tensors for test/convert to ensure falsely added constants cant pass

jaybdub committed:
more relu interfaces

jaybdub committed:
Merge pull request #90 from NVIDIA-AI-IOT/supported_elementwise

Supported elementwise

jaybdub committed:
fixed broadcast shapes, removed final_shapes

jaybdub committed:
removed all-dim test from max/min, since this cannot be supported by tensorrt

jaybdub committed:
removed leaf check

jaybdub committed:
added jetson xavier benchmark

jaybdub committed:
Merge pull request #91 from NVIDIA-AI-IOT/broadcast_fix

Broadcast fix

jaybdub committed:
added python2 serialization fix (bytes -> bytearray)

jaybdub committed:
added save_load test

jaybdub committed:
added save/load test

jaybdub committed:
Merge pull request #93 from NVIDIA-AI-IOT/py2_serialization_fix

Py2 serialization fix

jaybdub committed:
extra ignore

jaybdub committed:
Merge pull request #94 from NVIDIA-AI-IOT/ignore

extra ignore

jaybdub committed:
added leaky_relu

jaybdub committed:
added elu

jaybdub committed:
added selu, softsign, elu

jaybdub committed:
added softplus

jaybdub committed:
softplus, softsign

jaybdub committed:
Merge pull request #95 from NVIDIA-AI-IOT/activations_supported

Activations supported

jaybdub committed:
added network->dot graph util method

jaybdub committed:
added network->dot graph util method

jaybdub committed:
added network->dot graph util method

jaybdub committed:
added network->dot graph util method

jaybdub committed:
merged

jaybdub committed:
graphviz module prefix fix

jaybdub committed:
graphviz module prefix fix

jaybdub committed:
fixed Linear w.o bias

jaybdub committed:
Merge pull request #125 from jaybdub/linear_no_bias

fixed Linear w.o bias

jaybdub committed:
handle non-existance torch method with existing converter

jaybdub committed:
handle non-existance torch method with existing converter

jaybdub committed:
added profiling

jaybdub committed:
added profiling

jaybdub committed:
Merge branch 'network_visualization' of https://github.com/NVIDIA-AI-IOT/torch2trt into network_visualization

jaybdub committed:
Merge pull request #106 from NVIDIA-AI-IOT/network_visualization

Network visualization

jaybdub committed:
added dummy

jaybdub committed:
pass for Tensor.dim Tensor.size dummy converters

jaybdub committed:
incremented patch

jaybdub committed:
Merge pull request #138 from NVIDIA-AI-IOT/unsupported_methods

Unsupported methods

Commit:
added upsample code in converters

Commit:
added functional <--> nn.module binding for upsample

Commit:
made changes:

Commit:
working upsample with trt6

Commit:
added F.interpolate converter

Commit:
(WIP), trt version check

Commit:
fixed upsampling<->interpolate translation

Commit:
reverted changes to fix build

jaybdub committed:
renamed test nn_scale->upsample_scale_factor2

jaybdub committed:
Merge branch 'SrivastavaKshitij-kshitij/upsample_plugin_fix'

jaybdub committed:
added int8 support

jaybdub committed:
minor refactor/comment fix of DatasetCalibrator

jaybdub committed:
handle loose dataset types

jaybdub committed:
fixed int8 support for TensorRT 5.0

jaybdub committed:
fix int8_support python2

jaybdub committed:
refactor int8 buffer assignment using zip()

jaybdub committed:
increment vers

jaybdub committed:
Merge pull request #168 from NVIDIA-AI-IOT/int8_support

Int8 support

jaybdub committed:
added instance norm from track_stats=True

jaybdub committed:
instance norm

jaybdub committed:
optimize instance norm by merging sqrt inside scale layer

jaybdub committed:
Merge pull request #193 from NVIDIA-AI-IOT/instance_norm

Instance norm

jaybdub committed:
added converter / tests for torch.Tensor.__getitem__

jaybdub committed:
Merge pull request #195 from NVIDIA-AI-IOT/getitem

added converter / tests for torch.Tensor.__getitem__

jaybdub committed:
fixed Linear for N, *, H case

jaybdub committed:
Merge pull request #196 from NVIDIA-AI-IOT/linear_fix

fixed Linear for N, *, H case

jaybdub committed:
fixed permute for list args

jaybdub committed:
Merge pull request #206 from NVIDIA-AI-IOT/permute_listarg_fix

fixed permute for list args

jaybdub committed:
added support for adaptive_max_pool2d using regular pooling

jaybdub committed:
Merge pull request #207 from NVIDIA-AI-IOT/adaptive_max_pool2d

added support for adaptive_max_pool2d using regular pooling

SnowMasaya committed:
update clamp

SnowMasaya committed:
rm idea

jaybdub committed:
Merge pull request #224 from SnowMasaya/clamp

Clamp isolated min/max

oliver-batchelor committed:
Add 3d convolutions and interpolations

oliver-batchelor committed:
Update lots of ops

oliver-batchelor committed:
More opreations

oliver-batchelor committed:
Remove tools import

jaybdub committed:
plugin setup.py

jaybdub committed:
asdf

jaybdub committed:
interpolate torch serialization

jaybdub committed:
version handling

FrogKuma committed:
fix padding issue of 2d transposed conv.

Commit:
fixed trt version compatibility for torch2trt.py

Commit:
separated trt 7 ops

Commit:
added original ops

Commit:
reverting to default for trt 7

Commit:
changed value from 7 to 7.0

Commit:
added unit test for ConvTranspose2d and fixed an indentation error in parent PR

jaybdub committed:
added testcase for commonly used kernel size 4, stride 2, padding 1 convtranspose configuration

jaybdub committed:
Merge branch 'SrivastavaKshitij-transpose_conv2d'

SrivastavaKshitij committed:
Merge branch 'master' into new_trt_ops

jaybdub committed:
added enabled kwarg to register_converter and add_module_test

jaybdub committed:
moved trt7 ops to converters dir

jaybdub committed:
merged interpolate modules

jaybdub committed:
format init

jaybdub committed:
increment minor vers

jaybdub committed:
added --plugins to setup

jaybdub committed:
plugins disclaimer to readme

jaybdub committed:
disable interpolate if PyTorch < 1.3

jaybdub committed:
fixed interpolate plugin checking

jaybdub committed:
Merge pull request #332 from NVIDIA-AI-IOT/SrivastavaKshitij-new_trt_ops

Srivastava kshitij new trt ops

jaybdub committed:
allowed plugin loading to silent fail

jaybdub committed:
Merge pull request #337 from NVIDIA-AI-IOT/plugin_silentfail

allowed plugin loading to silent fail

jaybdub committed:
added ONNX conversion alternative to torch2trt

jaybdub committed:
Merge pull request #338 from NVIDIA-AI-IOT/onnx_alternative

added ONNX conversion alternative to torch2trt

vellamike committed:
Add requirements.txt

vellamike committed:
Added install information

Added information regarding installing dependencies

vellamike committed:
Update README.md

FIxed error in instrucitons

jaybdub committed:
readme tensorrt install tip

jaybdub committed:
Merge pull request #361 from NVIDIA-AI-IOT/vellamike-update-requirements

Vellamike update requirements

Commit:
added preliminary code for narrow

Commit:
narrow converter added successfully

jaybdub committed:
softmax neg dim

jaybdub committed:
Merge pull request #396 from NVIDIA-AI-IOT/softmax_negdim_fix

softmax neg dim

jaybdub committed:
Merge pull request #393 from SrivastavaKshitij/narrow_converter

Narrow converter

jaybdub committed:
added add_missing_trt_tensors and broadcast_trt_tensors

jaybdub committed:
bugfix and add

jaybdub committed:
add tests

jaybdub committed:
div

jaybdub committed:
dimfix

jaybdub committed:
dump converters script

jaybdub committed:
dump converters script

jaybdub committed:
docs

jaybdub committed:
docs

jaybdub committed:
changelog contributing

jaybdub committed:
changelog contributing

jaybdub committed:
docs

jaybdub committed:
docs

jaybdub committed:
typos

jaybdub committed:
typos

jaybdub committed:
docs

jaybdub committed:
docs

Chujingjun committed:
expose the batch size of INT8 calibration as parameter, since different
size may generate different accuracy loss.

jaybdub committed:
Merge pull request #397 from NVIDIA-AI-IOT/const_dim_fix

Const dim fix

jaybdub committed:
Merge pull request #398 from Chujingjun/master

expose the batch size of INT8 calibration as parameter

jaybdub committed:
docs

jaybdub committed:
docs build contrib

jaybdub committed:
build docs

jaybdub committed:
asdf

jaybdub committed:
docs merge

jaybdub committed:
merge fixes

jaybdub committed:
asdf

jaybdub committed:
mkdocs

jaybdub committed:
docs

jaybdub committed:
contrib plain md

jaybdub committed:
basic usage

jaybdub committed:
issues pathfix

jaybdub committed:
broken link

jaybdub committed:
release build docs

jaybdub committed:
Merge pull request #399 from jaybdub/docs

Docs

jaybdub committed:
Reduce precision docs (#400)

* reduced precision docs

* support matrix

* docs

jaybdub committed:
Update reduced_precision.md (#402)

jaybdub committed:
Update mkdocs.yml

jaybdub committed:
remove ndim referneces to support older pytorch versions (#415)

alsrgv committed:
Assert that inputs are contiguous (#418)

* Assert that inputs are contiguous

* Turn non-contiguous tensors into contiguous

* Add unit test

* Fix tabs

jaybdub committed:
Add layer names (#432)

* Auto-generate custom layer names

* fixed layer name count key

* updated changelog for adding layer names

Co-authored-by: Alex Sergeev <asergeev@maka-ars.com>

jaybdub committed:
Sandeepkumar skb groupnorm plugin (#437)

* added plugin for GroupNorm

Co-authored-by: sandeepkumar-skb <sandeep.skb@uber.com>

jaybdub committed:
Koenvandesande remove duplicate filenames (#448)

* Remove duplicate filenames which do not work on Windows by merging files

* Fix

* relu tests

Co-authored-by: Koen van de Sande <koen@keplervision.eu>

jaybdub committed:
allow direct method setting to support custom layers (#460)

jaybdub committed:
custom converter to changelog (#462)

jaybdub committed:
added expand converter (#487)

meremeev committed:
Additional converters for floordiv, mod, ne, and torch::tensor() operations (#505)

* Initioal version of ne, floordiv, mod and tensor converters. Extend ops for relu and sigmoid.

* Converters for floordiv, mod, ne, and torch::tensor() operations . Extend relu and sigmoid converters to Tensor methods.

* Update CHANGELOG.md

jaybdub committed:
added passing of torch2trt_kwargs to conversion context (#482)

* added passing of torch2trt_kwargs to conversion context

* added passing of torch2trt_kwargs to conversion context

jaybdub committed:
added filter to floordiv to only enable for pytorch 1.6+ (#511)

* added filter to floordiv to only enable for pytorch 1.6+

* enabled soft failure for missing torch method

jaybdub committed:
increment version to 0.2.0 (#517)

jaybdub committed:
Release push docs tagfix (#519)

* increment version to 0.2.0

* realse push docs tagfix

jaybdub committed:
Update README.md

jaybdub committed:
Create CLA.md

jaybdub committed:
Functional conv (#535)

* added conv_functional

* add Tensor flatten

* update changelog for functional conv / flatten

* add site to gitignore

jaybdub committed:
added contributors page

jaybdub committed:
Added John Welsh to CONTRIBUTORS.md

jaybdub committed:
I have read and agree to the Contributor License Agreement as written in the file CLA.md of this project.  Signed, John Welsh

jaybdub committed:
Merge pull request #550 from jaybdub/contributors

Contributors

Commit:
added silu converter

Commit:
functional linear converter

Commit:
added gelu converter

Commit:
added layer_norm converter

jaybdub committed:
Merge pull request #570 from jaybdub/linear_functional_converter

Linear functional converter

Commit:
added converter for torch.roll

jaybdub committed:
Merge pull request #574 from jaybdub/roll_converter

added converter for torch.roll

Commit:
fixed merged conflict

Commit:
recreating issue with view converter , bs >1

jaybdub commented:
Hi @SrivastavaKshitij ,

Thanks for pointing this out.

This likely has to do with this line

https://github.com/NVIDIA-AI-IOT/torch2trt/blob/817f937be44627f81d3efed9fafd4f8a905e5916/torch2trt/torch2trt.py#L517

When we pass the example data into the model, we do it with batch size 1.

From my understanding, converters shouldn't modify the batch dimension, so this would be acceptable, but perhaps I'm missing something.

Let me know if this helps.

Best,
John

Commit:
added some print statements for debugging

SrivastavaKshitij commented:
Hey John, 

Did some more experiments. 

when I changed line 517 from 

`inputs = [tensor.clone()[0:1] for tensor in inputs]`

to 

`inputs = [tensor.clone() for tensor in inputs]`

I was able to see the batch dimension. So that's good. However, 

https://github.com/NVIDIA-AI-IOT/torch2trt/blob/817f937be44627f81d3efed9fafd4f8a905e5916/torch2trt/torch2trt.py#L153

removes the batch dimension again and I dont know why is that. What do you think ? 

I added a print statement of the tensor shape before and after line 157 and this is what I got 

```
torch.Size([2, 3, 3, 3]) 
(3, 3, 3)
```

Somehow, batch dim is removed. 

Now this may not have impacted other ops but it will impact some of the ops such as view because the volume of the tensor will not match 

```
[TensorRT] ERROR: [SHUFFLE #1] torch.Tensor.view(tensor(shape=[2, 3, 3, 3], dtype=torch.float32), 1, -1): volume mismatch. Input dimensions [3,3,3] have volume 27 and output dimensions [54] have volume 54.
```

jaybdub commented on torch2trt/converters/view.py:
```diff
@@ -45,13 +49,13 @@ def forward(self, x):
+#@add_module_test(torch.float32, torch.device('cuda'), [(1, 3, 3)])
+@add_module_test(torch.float32, torch.device('cuda'), [(2, 3, 3, 3)],max_batch_size=3)
 def test_view_1d():
     return View(1, -1)
```
It looks like the test case is hard-fixing the batch dimension to 1 here.  

In general, TensorRT engines are broadcast across the batch dimension, so operations that change the batch dimension aren't permitted.

Perhaps adding a test case with View(2, ...) would work for batch size to.  Or maybe even View(-1, ...) with other dimensions specified in full.

jaybdub replied:
It looks like the test case is hard-fixing the batch dimension to 1 here.  

In general, TensorRT engines are broadcast across the batch dimension, so operations that change the batch dimension aren't permitted.

Perhaps adding a test case with View(2, ...) would work for batch size to.  Or maybe even View(-1, ...) with other dimensions specified in full.
